### DISCUSSION TITLE
fab format question

### DISCUSSION THREAD

**Discussion by indra098124** [AMReX-Astro/MAESTROeX#473] [Q&A] [answered]
I see that when writing the checkpoint file we set fab format to native and then after writing the checkpoint we set it back to the previous format (specified by the user for e.g. single precision). Is there a data loss in the conversion? Can you point me to where do we ensure that the calculations are still being done in double precision? 

Thank you

**Comment by zingale** [comment, marked as answer]
We will always output the checkpoint files in whatever format the host computer is using, at full precision.  The fab format will allow you to swap the endianness of the plotfile or output it in reduced (i.e. 32-bit instead of 64-bit) precision.  

In some setups you'll see:

```
fab.format = NATIVE_32
```

this makes plotfiles single precision, mainly just to save space.  Most output tools can manage the byte-swapping if the endianness of the machine you are doing analysis on is different than the one that you are doing the simulations on, so you don't need to worry about that generally.

**Reply by indra098124** [reply]
Thanks @zingale. I looked at the following section of the code where first we set the FArrayBox to the NATIVE format and then after writing the checkpoint we convert the FArrayBox to the format set by the user. I assume this is because when AMReX checks for fab.format it globally sets FArrayBox to the user specified format, this is why the following section of the code needed (please correct me if I am wrong). How in MAESTROeX (or in general) it is ensured that there is no data loss? Or fab.format is only for writing, setFormat does not convert the precision of the data in memory?


 // In checkpoint files, always write out FABs in NATIVE format
    FABio::Format thePrevFormat = FArrayBox::getFormat();

    FArrayBox::setFormat(FABio::FAB_NATIVE);

    // write the MultiFab data to, e.g., chk00010/Level_0/
    for (int lev = 0; lev <= finest_level; ++lev) {
        VisMF::Write(snew[lev], amrex::MultiFabFileFullPrefix(
@@ -113,6 +118,9 @@ void Maestro::WriteCheckPoint(int step) {
#endif
    }

    // Restore the previous FAB format.
    FArrayBox::setFormat(thePrevFormat);

    // write out the cell-centered base state
    if (ParallelDescriptor::IOProcessor()) {
        std::ofstream BaseCCFile;

**Reply by indra098124** [reply]
I think I found the answer in AMReX_FArrayBox.H. Thank you
